File: python/src/buble_ai/resources/generations.py
from __future__ import annotations

import asyncio
import time
from typing import Any, Dict, List, Optional

from .._errors import BubleCanceledError, BubleGenerationError, BubleTimeoutError
from .._http import AsyncBubleHTTPClient, BubleHTTPClient
from ..types.generations import GenerationTaskEnvelope

FORBIDDEN_GENERATION_FIELDS = {
    "input",
    "options",
    "scene",
    "sub_mode_id",
    "provider",
    "mediaType",
    "media_type",
    "images",
    "image_input",
    "video_input",
    "audio_input",
}

TERMINAL_STATUSES = {"success", "failed", "canceled"}


def _compact_body(body: Dict[str, Any]) -> Dict[str, Any]:
    return {key: value for key, value in body.items() if value is not None}


def _assert_public_generation_body(body: Dict[str, Any]) -> None:
    forbidden = FORBIDDEN_GENERATION_FIELDS.intersection(body.keys())
    if forbidden:
        field = sorted(forbidden)[0]
        raise ValueError(f'Field "{field}" is an internal Buble field and is not supported by the public API.')


class GenerationsResource:
    def __init__(self, http: BubleHTTPClient) -> None:
        self._http = http

    def create(
        self,
        *,
        model: str,
        mode: Optional[str] = None,
        prompt: Optional[str] = None,
        image_urls: Optional[List[str]] = None,
        start_frame: Optional[str] = None,
        end_frame: Optional[str] = None,
        video_urls: Optional[List[str]] = None,
        audio_urls: Optional[List[str]] = None,
        is_public: Optional[bool] = None,
        copy_protected: Optional[bool] = None,
        timeout: Optional[float] = None,
        **params: Any,
    ) -> GenerationTaskEnvelope:
        body = _compact_body(
            {
                "model": model,
                "mode": mode,
                "prompt": prompt,
                "image_urls": image_urls,
                "start_frame": start_frame,
                "end_frame": end_frame,
                "video_urls": video_urls,
                "audio_urls": audio_urls,
                "is_public": is_public,
                "copy_protected": copy_protected,
                **params,
            }
        )
        _assert_public_generation_body(body)
        return self._http.request("POST", "/api/v1/generations", json=body, timeout=timeout)

    def retrieve(self, task_id: str, *, timeout: Optional[float] = None) -> GenerationTaskEnvelope:
        return self._http.request("GET", f"/api/v1/generations/{task_id}", timeout=timeout)

    def wait(
        self,
        task_id: str,
        *,
        interval: float = 2.0,
        timeout: float = 600.0,
        throw_on_failed: bool = True,
        throw_on_canceled: bool = True,
    ) -> GenerationTaskEnvelope:
        start = time.monotonic()
        while True:
            if time.monotonic() - start > timeout:
                raise BubleTimeoutError(f"Generation {task_id} did not finish within {timeout}s.", timeout=timeout)

            envelope = self.retrieve(task_id)
            task = envelope["data"]
            status = task.get("status")
            if status in TERMINAL_STATUSES:
                if status == "failed" and throw_on_failed:
                    error = task.get("error") or {}
                    raise BubleGenerationError(error.get("message") or "Generation failed.", task=task)
                if status == "canceled" and throw_on_canceled:
                    raise BubleCanceledError(f"Generation {task_id} was canceled.", task=task)
                return envelope

            time.sleep(interval)


class AsyncGenerationsResource:
    def __init__(self, http: AsyncBubleHTTPClient) -> None:
        self._http = http

    async def create(
        self,
        *,
        model: str,
        mode: Optional[str] = None,
        prompt: Optional[str] = None,
        image_urls: Optional[List[str]] = None,
        start_frame: Optional[str] = None,
        end_frame: Optional[str] = None,
        video_urls: Optional[List[str]] = None,
        audio_urls: Optional[List[str]] = None,
        is_public: Optional[bool] = None,
        copy_protected: Optional[bool] = None,
        timeout: Optional[float] = None,
        **params: Any,
    ) -> GenerationTaskEnvelope:
        body = _compact_body(
            {
                "model": model,
                "mode": mode,
                "prompt": prompt,
                "image_urls": image_urls,
                "start_frame": start_frame,
                "end_frame": end_frame,
                "video_urls": video_urls,
                "audio_urls": audio_urls,
                "is_public": is_public,
                "copy_protected": copy_protected,
                **params,
            }
        )
        _assert_public_generation_body(body)
        return await self._http.request("POST", "/api/v1/generations", json=body, timeout=timeout)

    async def retrieve(self, task_id: str, *, timeout: Optional[float] = None) -> GenerationTaskEnvelope:
        return await self._http.request("GET", f"/api/v1/generations/{task_id}", timeout=timeout)

    async def wait(
        self,
        task_id: str,
        *,
        interval: float = 2.0,
        timeout: float = 600.0,
        throw_on_failed: bool = True,
        throw_on_canceled: bool = True,
    ) -> GenerationTaskEnvelope:
        start = time.monotonic()
        while True:
            if time.monotonic() - start > timeout:
                raise BubleTimeoutError(f"Generation {task_id} did not finish within {timeout}s.", timeout=timeout)

            envelope = await self.retrieve(task_id)
            task = envelope["data"]
            status = task.get("status")
            if status in TERMINAL_STATUSES:
                if status == "failed" and throw_on_failed:
                    error = task.get("error") or {}
                    raise BubleGenerationError(error.get("message") or "Generation failed.", task=task)
                if status == "canceled" and throw_on_canceled:
                    raise BubleCanceledError(f"Generation {task_id} was canceled.", task=task)
                return envelope

            await asyncio.sleep(interval)

File: flutter/lib/src/generations.dart
import 'dart:async';

import 'errors.dart';
import 'http_client.dart';
import 'types.dart';
import 'url_coding.dart';

const _defaultWaitInterval = Duration(seconds: 2);
const _defaultWaitTimeout = Duration(minutes: 10);

final class CreateGenerationRequest {
  CreateGenerationRequest({
    required this.model,
    this.mode,
    this.prompt,
    List<String>? imageUrls,
    this.startFrame,
    this.endFrame,
    List<String>? videoUrls,
    List<String>? audioUrls,
    this.isPublic,
    this.copyProtected,
    Map<String, Object?>? parameters,
  })  : imageUrls = imageUrls ?? const [],
        videoUrls = videoUrls ?? const [],
        audioUrls = audioUrls ?? const [],
        parameters = parameters ?? const {};

  final String model;
  final String? mode;
  final String? prompt;
  final List<String> imageUrls;
  final String? startFrame;
  final String? endFrame;
  final List<String> videoUrls;
  final List<String> audioUrls;
  final bool? isPublic;
  final bool? copyProtected;
  final Map<String, Object?> parameters;

  CreateGenerationRequest withParam(String key, Object? value) {
    assertSupportedGenerationField(key);
    return copyWith(parameters: {...parameters, key: value});
  }

  CreateGenerationRequest withParams(Map<String, Object?> values) {
    for (final key in values.keys) {
      assertSupportedGenerationField(key);
    }
    return copyWith(parameters: {...parameters, ...values});
  }

  CreateGenerationRequest copyWith({
    String? model,
    String? mode,
    String? prompt,
    List<String>? imageUrls,
    String? startFrame,
    String? endFrame,
    List<String>? videoUrls,
    List<String>? audioUrls,
    bool? isPublic,
    bool? copyProtected,
    Map<String, Object?>? parameters,
  }) {
    return CreateGenerationRequest(
      model: model ?? this.model,
      mode: mode ?? this.mode,
      prompt: prompt ?? this.prompt,
      imageUrls: imageUrls ?? this.imageUrls,
      startFrame: startFrame ?? this.startFrame,
      endFrame: endFrame ?? this.endFrame,
      videoUrls: videoUrls ?? this.videoUrls,
      audioUrls: audioUrls ?? this.audioUrls,
      isPublic: isPublic ?? this.isPublic,
      copyProtected: copyProtected ?? this.copyProtected,
      parameters: parameters ?? this.parameters,
    );
  }

  JsonMap toJson() {
    final body = <String, Object?>{
      'model': model,
      if (mode != null && mode!.isNotEmpty) 'mode': mode,
      if (prompt != null && prompt!.isNotEmpty) 'prompt': prompt,
      if (imageUrls.isNotEmpty) 'image_urls': imageUrls,
      if (startFrame != null && startFrame!.isNotEmpty)
        'start_frame': startFrame,
      if (endFrame != null && endFrame!.isNotEmpty) 'end_frame': endFrame,
      if (videoUrls.isNotEmpty) 'video_urls': videoUrls,
      if (audioUrls.isNotEmpty) 'audio_urls': audioUrls,
      if (isPublic != null) 'is_public': isPublic,
      if (copyProtected != null) 'copy_protected': copyProtected,
      ...parameters,
    };
    for (final key in body.keys) {
      assertSupportedGenerationField(key);
    }
    return body;
  }
}

final class WaitOptions {
  const WaitOptions({
    this.interval = _defaultWaitInterval,
    this.timeout = _defaultWaitTimeout,
    this.throwOnFailed = true,
    this.throwOnCanceled = true,
  });

  final Duration interval;
  final Duration timeout;
  final bool throwOnFailed;
  final bool throwOnCanceled;
}

final class GenerationsResource {
  const GenerationsResource(this._http);

  final BubleHttpClient _http;

  Future<Envelope<GenerationTask>> create(CreateGenerationRequest request) {
    return _http.post(
      '/api/v1/generations',
      body: request.toJson(),
      decoder: (json) =>
          Envelope(data: GenerationTask.fromJson(asJsonMap(json['data']))),
    );
  }

  Future<Envelope<GenerationTask>> retrieve(String id) {
    return _http.get(
      '/api/v1/generations/${encodePathSegment(id)}',
      decoder: (json) =>
          Envelope(data: GenerationTask.fromJson(asJsonMap(json['data']))),
    );
  }

  Future<Envelope<GenerationTask>> wait(
    String id, {
    WaitOptions options = const WaitOptions(),
  }) async {
    final deadline = DateTime.now().add(options.timeout);
    while (true) {
      final envelope = await retrieve(id);
      final task = envelope.data;
      if (task.status.isTerminal) {
        if (task.status == TaskStatus.failed && options.throwOnFailed) {
          throw GenerationFailedException(task);
        }
        if (task.status == TaskStatus.canceled && options.throwOnCanceled) {
          throw GenerationCanceledException(task);
        }
        return envelope;
      }

      if (DateTime.now().isAfter(deadline)) {
        throw BubleTimeoutException(
          'Generation $id did not finish within ${options.timeout}.',
          options.timeout,
        );
      }
      await Future<void>.delayed(options.interval);
    }
  }
}

void assertSupportedGenerationField(String field) {
  if (_forbiddenGenerationFields.contains(field)) {
    throw UnsupportedGenerationFieldException(field);
  }
}

const _forbiddenGenerationFields = {
  'input',
  'options',
  'scene',
  'sub_mode_id',
  'subModeId',
  'provider',
  'mediaType',
  'media_type',
  'images',
  'image_input',
  'video_input',
  'audio_input',
};
